[Refactor] Implement a `CoreGraph` class according to the design discussion. (1)

daehyun99 opened the pull request:
### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #2383

### List of changes to the codebase in this pull request
- Change AncestralBase.py file name to base.py.
- Change test_AncestralBase.py file name to test_base.py.
- Change `AncestralBase` class name to `CoreGraph`.
- Modifiy the comment for the `__eq__` method of `CoreGraph`.
- Modifiy the code for the `copy` method of `CoreGraph`.

daehyun99 committed:
Refactor: Change `AncestralBase` to `CoreGraph`(#2383)

- Change AncestralBase.py file name to base.py.
- Change test_AncestralBase.py file name to test_base.py.
- Change `AncestralBase` class name to `CoreGraph`.

daehyun99 committed:
Refactor: Update the code and comment for the method of `CoreGraph`. (#2383)

- Modifie the comment for the `__eq__` method.
- Modifie the code for the `copy` method.

daehyun99 committed:
Merge branch 'dev' of https://github.com/daehyun99/pgmpy into Refactor/2383/CoreGraph

ankurankan committed:
Merge branch 'dev' into Refactor/2383/CoreGraph

ankurankan commented:
@daehyun99 I have merged `dev` into your branch to fix the failing tests. Please make sure to do a `git pull` before pushing changes.

daehyun99 commented:
Thanks @ankurankan. I missed that.

daehyun99 committed:
Refactor: Change `nx.Graph` to `nx.MultiGraph` for `CoreGraph`. (#2392)

- This is a temporary commit to check the full Github Action CI test results.
- This is a temporary commit to understand the difference between `nx.Graph` and `nx.MultiGraph`.

daehyun99 committed:
Refactor: Implement abstract methods for `CoreGraph`. (#2383)

- Override `nx.MultiGraph.is_directed`.
- Override `nx.MultiGraph.is_multigraph`.

daehyun99 committed:
Fix: fix pre-commit (#2383)

ankurankan commented:
@daehyun99 I think it might be a better idea to create a new class for `CoreGraph` instead of replacing `AncestralBase`. That would allow us to develop the `CoreGraph` class separately without touching the other child classes. And once the `CoreGraph` is ready we can start replacing the `AncestralBase` class (the third sub-issue).

daehyun99 commented:
@ankurankan 
Yes, I agree too.

Furthermore, I think the base.py file should be renamed to BaseGraph.py or CoreGraph.py because a file with the same name already exists in the path pgmpy/sampling/base.py.

Thanks.

daehyun99 commented:
@ankurankan 
I think it would be better to create a new Pull Request rather than reverting this commit. I will continue this in #2402.

ankurankan commented:
> because a file with the same name already exists in the path pgmpy/sampling/base.py.

I think having a `base.py` file is fine. Most modules in pgmpy have a `base.py` file that implements the base functionality of that module.
